File: gigworkers/gig_workers/report/service_rate_history/service_rate_history.js
// Copyright (c) 2026, Jenifar and contributors
// For license information, please see license.txt

frappe.query_reports["Service Rate History"] = {
	filters: [
		{
			fieldname: "service",
			label:     __("Service"),
			fieldtype: "Link",
			options:   "Service",
		},
		{
			fieldname: "category",
			label:     __("Category"),
			fieldtype: "Link",
			options:   "Service Category",
		},
		{
			fieldname: "vehicle_type",
			label:     __("Vehicle Type"),
			fieldtype: "Link",
			options:   "Vehicle Type",
		},
		{
			fieldname: "status",
			label:     __("Status"),
			fieldtype: "Select",
			options:   "\nActive\nScheduled\nExpired",
		},
	],

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (!data) return value;

		if (column.fieldname === "status") {
			const cfg = {
				Active:    { color: "#155724", bg: "#d4edda", border: "#c3e6cb" },
				Scheduled: { color: "#856404", bg: "#fff3cd", border: "#ffeeba" },
				Expired:   { color: "#721c24", bg: "#f8d7da", border: "#f5c6cb" },
			}[data.status];
			if (cfg) {
				return `<span style="
					background:${cfg.bg};color:${cfg.color};
					border:1px solid ${cfg.border};
					padding:2px 10px;border-radius:12px;
					font-size:12px;font-weight:600;
					display:inline-block;
				">${data.status}</span>`;
			}
		}

		if (column.fieldname === "welfare_percentage_" && data.welfare_percentage_ != null) {
			return `<span style="font-weight:600;">${data.welfare_percentage_}%</span>`;
		}

		return value;
	},

	after_refresh() {
		// Add padding and row height to the report table
		const style = document.getElementById("srhist-style");
		if (style) return;

		const el = document.createElement("style");
		el.id = "srhist-style";
		el.textContent = `
			.dt-cell__content { padding: 8px 12px !important; }
			.dt-row { height: 44px !important; }
			.dt-header .dt-cell__content {
				padding: 10px 12px !important;
				font-weight: 700 !important;
				font-size: 12px !important;
				text-transform: uppercase;
				letter-spacing: .4px;
				color: #555 !important;
				background: #f8f9fa;
			}
			.dt-row:hover .dt-cell { background: #f0f4ff !important; }
		`;
		document.head.appendChild(el);
	},
};

File: gigworkers/gig_workers/report/service_rate_history/service_rate_history.py
# Copyright (c) 2026, Jenifar and contributors
# For license information, please see license.txt

import frappe

# Status is computed live from each row's own dates rather than trusting
# log.status, which is only a snapshot of what the status was when the row
# was logged and never gets updated afterwards.
LIVE_STATUS_SQL = """
	CASE
		WHEN log.effective_end_date IS NOT NULL AND log.effective_end_date < CURDATE() THEN 'Expired'
		WHEN log.effective_start_date IS NOT NULL AND log.effective_start_date > CURDATE() THEN 'Scheduled'
		ELSE 'Active'
	END
"""


def execute(filters=None):
	filters = filters or {}
	columns = get_columns()
	data    = get_data(filters)
	return columns, data


def get_columns():
	return [
		{
			"fieldname": "service",
			"label":     "Service",
			"fieldtype": "Link",
			"options":   "Service",
			"width":     110,
		},
		{
			"fieldname": "category_name",
			"label":     "Category",
			"fieldtype": "Data",
			"width":     150,
		},
		{
			"fieldname": "vehicle_type_name",
			"label":     "Vehicle Type",
			"fieldtype": "Data",
			"width":     130,
		},
		{
			"fieldname": "welfare_percentage_",
			"label":     "Welfare %",
			"fieldtype": "Float",
			"width":     110,
		},
		{
			"fieldname": "welfare_cap",
			"label":     "Welfare Cap (₹)",
			"fieldtype": "Currency",
			"width":     150,
		},
		{
			"fieldname": "effective_start_date",
			"label":     "Start Date",
			"fieldtype": "Date",
			"width":     120,
		},
		{
			"fieldname": "effective_end_date",
			"label":     "End Date",
			"fieldtype": "Date",
			"width":     120,
		},
		{
			"fieldname": "status",
			"label":     "Status",
			"fieldtype": "Data",
			"width":     110,
		},
	]


def get_data(filters):
	conditions = []
	values     = {}

	if filters.get("service"):
		conditions.append("log.parent = %(service)s")
		values["service"] = filters["service"]

	if filters.get("status"):
		conditions.append(LIVE_STATUS_SQL + " = %(status)s")
		values["status"] = filters["status"]

	if filters.get("category"):
		conditions.append("svc.category = %(category)s")
		values["category"] = filters["category"]

	if filters.get("vehicle_type"):
		conditions.append("svc.vehicle_type = %(vehicle_type)s")
		values["vehicle_type"] = filters["vehicle_type"]

	where = ("WHERE " + " AND ".join(conditions)) if conditions else ""

	return frappe.db.sql("""
		SELECT
			log.parent                          AS service,
			COALESCE(sc.category_name, svc.category)   AS category_name,
			COALESCE(vt.vehicle_type, svc.vehicle_type) AS vehicle_type_name,
			log.welfare_percentage_,
			log.welfare_cap,
			log.effective_start_date,
			log.effective_end_date,
			{live_status} AS status
		FROM `tabService Rate Log` log
		JOIN `tabService` svc ON svc.name = log.parent
		LEFT JOIN `tabService Category` sc ON sc.name = svc.category
		LEFT JOIN `tabVehicle Type` vt ON vt.name = svc.vehicle_type
		{where}
		ORDER BY log.parent, log.creation DESC
	""".format(live_status=LIVE_STATUS_SQL, where=where), values, as_dict=True)
